src/Discord.Net.Core/Net/IRequest.cs

```diff
@@ -0,0 +1,14 @@
+using System;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace Discord.Net
+{
+    public interface IRequest
+    {
+        DateTimeOffset? TimeoutAt { get; }
+        TaskCompletionSource<Stream> Promise { get; }
```

Comment:
I don't think we should expose the stream. Sounds like a gross implementation detail.

Reply:
So, also change the implementations' properties to be internal for `Promise`?